doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,214 @@
+which securely store sensitive values.
+
+## Design
+
+### Secret SQL syntax and semantics
+
+Four new SQL statements will be introduced with the following grammar to
+create, alter, drop, and list secrets, respectively:
+
+```
+create_secret_stmt ::=
+  CREATE SECRET [IF NOT EXISTS] <name> AS <value>
+
+alter_secret_stmt ::=
+  ALTER SECRET <name> AS <value>
+
+drop_secret_stmt ::=
+  DROP SECRET [IF EXISTS] <name> [{ RESTRICT | CASCADE }]
+
+show_secrets_stmt: SHOW SECRETS [FROM <schema>] [{ LIKE 'pattern' | WHERE <expr> }]
+
+<name> ::= <qualified-identifier>
+<value> ::= <scalar-expression>
+```
+
+**Initially, all of these statements will be available in experimental mode
+only.**
+
+The name of a secret is a qualified identifier. Secrets live in the standard
+database and schema hierarchy alongside tables, views, sources, etc.
+
+The `CREATE SECRET` statement creates a new secret with the specified contents.
+It returns an error if a secret with the specified name already exists unless
+the `IF NOT EXISTS` clause is present. The `<value>` expression may not
+reference any relations and must be implicitly castable to `bytea`.
+
+The `ALTER SECRET` statement changes the contents of an existing secret.
+
+The `DROP SECRET` statement drops an existing secret. It returns an error
+if a secret with the specified name does not exist unless the `IF EXIST`
+clause is present. Dropping a secret that is in use by a `CONNECTOR` is not
+permitted unless `CASCADE` is present.
+
+If the [secrets controller](#secrets-controller) in use supports multi-operation
+atomicity, a SQL transaction can contain any number of secret DDL statements;
+otherwise a SQL transaction can contain only one secret DDL statement. In either
+case, the SQL transaction must not contain any other type of statements. The
+secret DDL statements are applied atomically to the secrets controller if and
+only if the transaction commits.
+
+Note that while the catalog updates to a secret apply immediately, changes to
+the contents of the secret do not propagate to downstream consumers of the
+secret on any particular timeline.
+
+The `SHOW SECRETS` statement lists the *names* of the secrets in the specified
+schema, or the first schema in the search path if no schema is explicitly
+specified, optionally filtered by the provided `LIKE` pattern or `WHERE`
+expression. The filter clauses which work analogously to the same clauses in the
+`SHOW DATABASES` statement.
+
+Note that there is intentionally no way to view the contents of secrets.
+
+### System catalog changes
+
+A new `mz_secrets` table with the following structure will describe the
+available secrets in the system:
+
+Field       | Type      | Meaning
+------------|-----------|--------
+`id`        | `bigint`  | The ID of the secret.
+`schema_id` | `bigint`  | The ID of the schema to which the secret belongs.
+`name`      | `text`    | The name of the secret.
```

Comment:
the `name` here is a filename, right? If all secrets map to the same location on disk, they all have to be unique. How can we enforce uniqueness across databases/schemas and names?

Reply:
No, the name here is the name of the secret object in the schema.

>  If all secrets map to the same location on disk, they all have to be unique. How can we enforce uniqueness across databases/schemas and names?

Secrets are stored on disk with the global ID as the filename, as described in:

> All secrets controllers expose the contents of secrets to the [STORAGE](https://github.com/benesch/materialize/blob/secrets-design/doc/developer/platform/architecture-db.md#STORAGE) and [COMPUTE](https://github.com/benesch/materialize/blob/secrets-design/doc/developer/platform/architecture-db.md#COMPUTE) layer by way of a directory that contains one file per secret, where the name of the file is the secret's ID and the contents of the file are the secret's contents. For example, if IDs u12 and u36 are secrets, the directory would look like this... [snip]